# 1. Some customers didn't write a review. But why are  they happy or dissatisfied?

In [1]:
import pandas as pd
pd.set_option('max_colwidth',200)
reviews = pd.read_csv('order_reviews.csv')
reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                                                                                 review_comment_message  \
0                                                                                                   NaN   
1                                                                                                   NaN   
2                                                                                                   NaN   
3                                                                 Recebi bem antes do prazo estipulado.   
4  Parabéns lojas lannister adorei comprar pela Internet seguro e prático Parabéns a todos feliz Páscoa   

  review_creation_date review_answer_timestamp  
0  2018-01-18 00:00:00     2018-01-18 21:46:59  
1  2018-03-10 00:00:00     2018-03-11 03:05:13  
2  2018-02-17 00:00:00     2018-02-18 14:36:24  
3  2017-04-21 00:00:00     2017-04-21 22:02:06  
4  2018-03-01 00:00:00     2018-03-02 10:26:53

In [2]:
reviews.shape,reviews.drop_duplicates().shape

((100000, 7), (100000, 7))

we got 100000 examples total with no duplicate examples

In [3]:
# get to know types of each columns
reviews.dtypes

review_id                  object
order_id                   object
review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
dtype: object

In [4]:
reviews['review_score'].unique()

array([4, 5, 1, 3, 2])

In [5]:
reviews['review_score'].describe()

count    100000.000000
mean          4.070890
std           1.359663
min           1.000000
25%           4.000000
50%           5.000000
75%           5.000000
max           5.000000
Name: review_score, dtype: float64

In [6]:
reviews.groupby('review_score')['review_id'].count()

review_score
1    11858
2     3235
3     8287
4    19200
5    57420
Name: review_id, dtype: int64

In [7]:
reviews.groupby('review_score')['order_id'].count()

review_score
1    11858
2     3235
3     8287
4    19200
5    57420
Name: order_id, dtype: int64

In [8]:
reviews.sample(10)

review_id                          order_id  \
16266  b5da8c372f190911d1218077c3e432cf  9a965cd710da5125076afbd641624600   
75358  072133dc43d0fd76a78b665fa716b140  37e1d9b8fc877b2c60bf51d2c4d8dbc6   
58637  63272ba8a58647be696ae623e380f811  a10dfacbf1c8bc9c9e9591f60102bdef   
83040  dc5cdb59f1ee3eba8daa43c6c3486353  30647325eed2f0ac046e27a3cb6b76fc   
90499  34284dd3e90baa1e4c25c26fc380b87b  48c3b4aae43bbf8e6929c347112cba89   
59668  b4a147c05372e3ffa8aac9da627950e1  34a9f5d9dc692eef97911d47564b1af1   
96006  ac97c5912defd76672146f1cc8d11498  1fb40f920a832437e543e136c7a8b050   
18878  c3682d2306d35a08c131aa228bb3ef2e  c870c1f3a9ba62194d1060c0accb18dd   
51510  25b5083f478c1c2c328861ed577331ca  b19b0d1dcf07ad7d58c75f45e9744ba5   
17743  d6de7b5b09aec7945e550724ca520b37  3e555879d29880ec5bc189b44a6d77fa   

       review_score review_comment_title  \
16266             5                  NaN   
75358             5                  NaN   
58637             1              Pessimo   
83040             5                  NaN   
90499             5                  NaN   
59668             3                  NaN   
96006             5                  NaN   
18878             4                  NaN   
51510             5                  NaN   
17743             5                  NaN   

                                                                        review_comment_message  \
16266                                                                                      NaN   
75358                                                                     Bom prazo de entrega   
58637                                            Nunca mais vou comprar pelas lojas lannister    
83040  foi otimo ser atendida por vcs ,veio no prazo certo e é um otimo produto muito obrigado   
90499                                                                                      NaN   
59668                                                                                      NaN   
96006                                                                                      NaN   
18878                                                                                      NaN   
51510   Meu filho amou a mochila... Perfeita..e melhor ainda chegou justo no aniversário dele.   
17743                                                                                      NaN   

      review_creation_date review_answer_timestamp  
16266  2017-05-24 00:00:00     2017-05-27 20:58:18  
75358  2018-01-10 00:00:00     2018-01-13 02:13:45  
58637  2018-08-15 00:00:00     2018-08-15 15:23:00  
83040  2017-03-11 00:00:00     2017-03-22 10:48:53  
90499  2017-12-12 00:00:00     2017-12-15 12:09:46  
59668  2017-06-20 00:00:00     2017-06-21 00:12:42  
96006  2018-07-27 00:00:00     2018-07-28 03:14:11  
18878  2017-02-02 00:00:00     2017-02-04 08:19:08  
51510  2017-10-15 01:00:00     2017-10-15 19:38:22  
17743  2017-03-14 00:00:00     2017-03-15 07:26:45

let's take a look at 10 random sample examples.

we can see that both the 'review_comment_title'& 'review_comment_message' contains null values.
If we want to know why they are happy or dissatisfied,we need to take ['review_score','review_comment_title','review_comment_message','review_creation_date','review_answer_timestamp'] into consideration

In [9]:
reviews = reviews[['review_score','review_comment_title','review_comment_message','review_creation_date','review_answer_timestamp']].drop_duplicates().dropna()

In [10]:
reviews.count()

review_score               9944
review_comment_title       9944
review_comment_message     9944
review_creation_date       9944
review_answer_timestamp    9944
dtype: int64

we got a review dataset of 9944 examples with no NaN values

In [20]:
reviews.dtypes

review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
dtype: object

In [11]:
reviews.head()

review_score     review_comment_title  \
9              4                recomendo   
15             5          Super recomendo   
19             1  Não chegou meu produto    
22             5                    Ótimo   
34             4               Muito bom.   

                                                                                                                                                            review_comment_message  \
9   aparelho eficiente. no site a marca do aparelho esta impresso como 3desinfector e ao chegar esta com outro nome...atualizar com a marca correta uma vez que é o mesmo aparelho   
15                                                                                                                        Vendedor confiável, produto ok e entrega antes do prazo.   
19                                                                                                                                                                         Péssimo   
22                                                                                                                                                                    Loja nota 10   
34                                                                Recebi exatamente o que esperava. As demais encomendas de outros vendedores atrasaram, mas esta chegou no prazo.   

   review_creation_date review_answer_timestamp  
9   2018-05-22 00:00:00     2018-05-23 16:45:47  
15  2018-05-23 00:00:00     2018-05-24 03:00:01  
19  2018-08-15 00:00:00     2018-08-15 04:10:37  
22  2018-07-10 00:00:00     2018-07-11 14:10:25  
34  2018-06-07 00:00:00     2018-06-09 18:44:02

In [28]:
reviews.review_creation_date = pd.to_datetime(reviews.review_creation_date)
reviews.review_answer_timestamp = pd.to_datetime(reviews.review_answer_timestamp)
reviews['respond_time'] = (reviews.review_answer_timestamp - reviews.review_creation_date)                           

In [30]:
reviews.respond_time.describe()

count                       9944
mean      3 days 13:51:21.910901
std      12 days 00:26:42.135061
min              0 days 03:11:11
25%       0 days 23:54:49.500000
50%       1 days 17:39:33.500000
75%       3 days 09:23:05.250000
max            411 days 10:39:13
Name: respond_time, dtype: object

In [52]:
# let's take a sample of comment title with all the customers give 5 score(our happy customers)
reviews.groupby('review_score')['review_comment_title'].get_group(5).sample(10)

81360    super recomendo
90240    Entrega rápida 
57361          Excelente
34198          Recomendo
90329          excelente
39337              otimo
87598         Lindíssimo
11870          Recomendo
48515             Ótimo 
88275          Recomendo
Name: review_comment_title, dtype: object

In [53]:
# let's also take a sample of 10 of the unhappy customers who give 1 score
reviews.groupby('review_score')['review_comment_title'].get_group(1).sample(10)

59964    nao recebi o produto
89802     Produto veio errado
31103    Produto não entregue
65525     Produto com defeito
23992            NAO ENTREGUE
74183            Produto ruim
85813    Nem NF foi emitida!!
33813       QUERO MEU PRODUTO
65135     Cancelaram a compra
68541             Não gostei.
Name: review_comment_title, dtype: object

In [55]:
reviews.groupby('review_score')['review_comment_message'].get_group(5).sample(10)

96191                                                               Eu só compro neste site tenho segurança ok
78011                                                                                               Satisfação
51262                                 Muito bom! Compra tranquila, Entrega super rápida. Recomendo para todos!
15126                                                                                              Muito bom. 
97769                                                                                      Tecnologia japonesa
92847                                                                 Entregaram antes do prazo estipulado....
61405                                                     Produto de ótima qualidade chegou bem antes do prazo
10881                                                                                                   Ótimo 
40697                                                   muito bem foi entregue antes da data estão de parabens
9

In [56]:
reviews.groupby('review_score')['review_comment_message'].get_group(1).sample(10)

39468                                                                                                                                        Não recebi o produto nem retorno do questionamento sobre o mesmo...
86261                                                                                                                          Não foi o produto que comprei pelas lannister, recebi duas extensões de cabo HDI.
73866    Bem comprei esse quadro pela targaryen,e solicitei troca imediatamente,quadro parecido com o do anúncio no entanto apresenta impressão sem qualidade,sem foco imagem trêmula,o material da tela terr...
34891                     Fiquei muito triste pq pedi um relógio, não tinha cancelei meu pedido e mesmo assim mandaram outro.Detalhe já tinha dito que não queria outro modelo,resumindo pedi extorno de cartão.
3339                                                                                                                                                Produto não func